xarray/core/combine.py

```diff
@@ -456,107 +146,372 @@ def _combine_nd(combined_ids, concat_dims, data_vars='all',
-                                                         data_vars=data_vars,
-                                                         coords=coords,
-                                                         compat=compat)
+        combined_ids = _combine_all_along_first_dim(combined_ids,
+                                                    dim=concat_dim,
+                                                    data_vars=data_vars,
+                                                    coords=coords,
+                                                    compat=compat)
     combined_ds = list(combined_ids.values())[0]
     return combined_ds
 
 
-def _auto_combine_all_along_first_dim(combined_ids, dim, data_vars,
-                                      coords, compat):
+def _combine_all_along_first_dim(combined_ids, dim, data_vars, coords, compat):
+
     # Group into lines of datasets which must be combined along dim
     # need to sort by _new_tile_id first for groupby to work
     # TODO remove all these sorted OrderedDicts once python >= 3.6 only
     combined_ids = OrderedDict(sorted(combined_ids.items(), key=_new_tile_id))
     grouped = itertools.groupby(combined_ids.items(), key=_new_tile_id)
 
+    # Combine all of these datasets along dim
     new_combined_ids = {}
     for new_id, group in grouped:
         combined_ids = OrderedDict(sorted(group))
         datasets = combined_ids.values()
-        new_combined_ids[new_id] = _auto_combine_1d(datasets, dim, compat,
-                                                    data_vars, coords)
+        new_combined_ids[new_id] = _combine_1d(datasets, dim, compat,
+                                               data_vars, coords)
     return new_combined_ids
 
 
-def vars_as_keys(ds):
-    return tuple(sorted(ds))
-
+def _combine_1d(datasets, concat_dim, compat='no_conflicts', data_vars='all',
+                coords='different'):
+    """
+    Applies either concat or merge to 1D list of datasets depending on value
+    of concat_dim
+    """
 
-def _auto_combine_1d(datasets, concat_dim=_CONCAT_DIM_DEFAULT,
-                     compat='no_conflicts',
-                     data_vars='all', coords='different'):
-    # This is just the old auto_combine function (which only worked along 1D)
     if concat_dim is not None:
-        dim = None if concat_dim is _CONCAT_DIM_DEFAULT else concat_dim
-        sorted_datasets = sorted(datasets, key=vars_as_keys)
-        grouped_by_vars = itertools.groupby(sorted_datasets, key=vars_as_keys)
-        concatenated = [_auto_concat(list(ds_group), dim=dim,
-                                     data_vars=data_vars, coords=coords)
-                        for id, ds_group in grouped_by_vars]
+        try:
+            combined = concat(datasets, dim=concat_dim, data_vars=data_vars,
+                              coords=coords)
+        except ValueError as err:
+            if "encountered unexpected variable" in str(err):
```

Comment:
This is a sign that we may want to refactor `concat`, though we can definitely save that for later :). Generally string matching of errors can be somewhat error prone -- it's better to catch a dedicated exception type.